api/src/main/java/org/pmiops/workbench/db/dao/DataSetServiceImpl.java

```diff
@@ -216,97 +190,140 @@ public DataSet saveDataSet(
-                    + columnNames.getSourceConceptIdColumn()
-                    + conceptSetListQuery
-                    + ")");
+        queryBuilder.append(
+            formatWhereClause(conceptSetListQuery, " WHERE \n(",
+                domainConceptIdInfo.getStandardConceptIdColumn(), " OR \n",
+                domainConceptIdInfo.getSourceConceptIdColumn(),
+                conceptSetListQuery, ")"));
         if (!includesAllParticipants) {
-          query = query.concat(" \nAND (PERSON_ID IN (" + cohortQueries + "))");
+          queryBuilder
+              .append(" \nAND (PERSON_ID IN (")
+              .append(cohortQueries)
+              .append("))");
         }
       } else if (!includesAllParticipants) {
-        query = query.concat(" \nWHERE PERSON_ID IN (" + cohortQueries + ")");
+        queryBuilder
+            .append(" \nWHERE PERSON_ID IN (")
+            .append(cohortQueries)
+            .append(")");
       }
 
-      queryMap.put(query, cohortParameters);
-      QueryJobConfiguration queryJobConfiguration =
+      final String completeQuery = queryBuilder.toString();
+      final ImmutableMap<String, QueryParameterValue> cohortParameters = cohortParametersBuilder
+          .build();
+
+      final QueryJobConfiguration queryJobConfiguration =
           bigQueryService.filterBigQueryConfig(
-              QueryJobConfiguration.newBuilder(query)
+              QueryJobConfiguration.newBuilder(completeQuery)
                   .setNamedParameters(cohortParameters)
                   .setUseLegacySql(false)
                   .build());
-      dataSetUtil.put(d.toString(), queryJobConfiguration);
+      resultBuilder.put(domain.toString(), queryJobConfiguration);
     }
-    return dataSetUtil;
+    return resultBuilder.build();
+  }
+
+  private void validateSelections(DataSetRequest dataSet, boolean includesAllParticipants,
+      List<Cohort> cohortsSelected, List<ConceptSet> conceptSetsSelected) {
+    if ((cohortsSelected.size() == 0 && !includesAllParticipants)
+        || ((conceptSetsSelected.size() == 0
+                && dataSet.getPrePackagedConceptSet().equals(PrePackagedConceptSetEnum.NONE))
+            && dataSet.getValues().size() == 0)) {
+      throw new BadRequestException("Data Sets must include at least one cohort and concept.");
+    }
+  }
+
+  private static String formatWhereClause(String conceptSetListQuery, String s,
+      String standardConceptIdColumn, String s2, String sourceConceptIdColumn,
+      String conceptSetListQuery2, String s3) {
+    return s
+        + standardConceptIdColumn
+        + conceptSetListQuery
+        + s2
+        + sourceConceptIdColumn
+        + conceptSetListQuery2
+        + s3;
+  }
+
+  private void validateSelectedConceptSetsForDomain(List<ConceptSet> conceptSetsSelected,
+      Domain domain) {
+    conceptSetsSelected.stream()
+        .map(ConceptSet::getDomain)
+        .distinct()
+        .filter(cs -> domain != Domain.PERSON)
```

Comment:
To clarify, yes I suspect this wasn't intended, but it might still be load-bearing because of reasons I didn't want to chase down at the time. The nice thing about refactoring is that it makes bugs more obvious. But yeah, I can't think why you'd ever want to check a higher more broadly scoped condition in an inner loop or stream.

I think the moral is that we've just about extended this approach of hand-rolling SQL queries as far as we safely can. I want to sneak in work on https://precisionmedicineinitiative.atlassian.net/browse/RW-3466 at some point.

Reply:
I do think we could probably remove this. If we want to not run this on PERSON, we should do that as an if statement and then return if PERSON.

Reply:
Turns out the global filter's effects of making the stream degenerate were necessary, so I've tried to replicate their effects in this version. I've got one unit test on the person situation, and @NehaBroad has some more in the works from what I hear.